SimplenoteTests/NotesListControllerTests.swift

```diff
@@ -0,0 +1,495 @@
+import XCTest
+import SimplenoteFoundation
+@testable import Simplenote
+
+
+// MARK: - NotesListControllerTests
+//
+class NotesListControllerTests: XCTestCase {
```

Comment:
I think these tests are great as an integration tests, but I don't think strictly speaking we can call them unit tests. 

I will try to explain why I think this way. 

Right now we're testing not only the logic of `NoteListController`, but also it's collaborators:
* `ResultsController` and in the end CoreData that is used to fetch data
* `NoteListState` which is used to generate predicates

So along with `NoteListController` we also test that core data fetches the right entities and `NoteListState` produces correct predicates. But is it what we really want to test here? I would say no. I hope that Core Data is already tested and when provided with a correct predicate will return correct entities.

If we test `ResultsController` and `NoteListState` separately and we know that they work, then when testing `NoteListController` we can fully mock them and test that during certain code path certain methods are invoked.

Example of some tests I would write to test the behaviour of `NoteListController`:
* When we init `NoteListController`, then `onWillChangeContent` and `onDidChangeContent` is set in `ResultsController`
* When `ResultsController` calls `onWillChangeContent` / `onDidChangeContent`, then respective callbacks from `NoteListController` are invoked. The actual content `objectsChangeset` is not relevant here because `NoteListController` doesn't do anything with it, we just want to make sure that it is passed along.
* When filter is set on `NoteListController`, then we check that we get predicates from `NoteListState` and set them on `ResultsController`. Again for `NoteListController`  it is not really important which filter is chosen, because it's job is to take the predicates for this filter from the state and set it to `ResultsController`

I think this way we can improve by only testing behaviour of one object at a time (if possible).

What do you think? I can elaborate more on how I see this :-)